File: lms/nodejs/src/routes/analytics.js
const express = require('express');
const router = express.Router();

// Mock analytics database
const analytics = {};

// Record user activity
router.post('/track', (req, res) => {
  try {
    const { userId, courseId, action, timestamp } = req.body;
    const key = `${userId}-${courseId}`;

    if (!analytics[key]) {
      analytics[key] = {
        userId,
        courseId,
        activities: [],
        lastActive: timestamp
      };
    }

    analytics[key].activities.push({ action, timestamp });
    analytics[key].lastActive = timestamp;

    res.json({ message: 'Activity tracked' });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Get user progress
router.get('/progress/:userId/:courseId', (req, res) => {
  try {
    const key = `${req.params.userId}-${req.params.courseId}`;
    const data = analytics[key];

    if (!data) {
      return res.json({ userId: req.params.userId, courseId: req.params.courseId, progress: 0 });
    }

    res.json({
      userId: req.params.userId,
      courseId: req.params.courseId,
      totalActivities: data.activities.length,
      lastActive: data.lastActive
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Get course analytics
router.get('/course/:courseId', (req, res) => {
  try {
    const courseAnalytics = Object.values(analytics).filter(
      a => a.courseId == req.params.courseId
    );

    res.json({
      courseId: req.params.courseId,
      totalUsers: courseAnalytics.length,
      totalActivities: courseAnalytics.reduce((sum, a) => sum + a.activities.length, 0)
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;

File: lms/rust/src/handlers/analytics.rs
use actix_web::{web, HttpResponse};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct TrackActivityRequest {
    pub user_id: String,
    pub course_id: u32,
    pub action: String,
    pub timestamp: String,
}

pub async fn track_activity(req: web::Json<TrackActivityRequest>) -> HttpResponse {
    HttpResponse::Ok().json(serde_json::json!({
        "message": "Activity tracked"
    }))
}

pub async fn get_progress(path: web::Path<(String, u32)>) -> HttpResponse {
    let (user_id, course_id) = path.into_inner();
    HttpResponse::Ok().json(serde_json::json!({
        "user_id": user_id,
        "course_id": course_id,
        "progress": 0.0
    }))
}

pub async fn get_course_analytics(course_id: web::Path<u32>) -> HttpResponse {
    HttpResponse::Ok().json(serde_json::json!({
        "course_id": course_id.into_inner(),
        "total_users": 0,
        "total_activities": 0
    }))
}
